File: Java/SpaceVIL/spacevil/src/main/java/com/spvessel/spacevil/EventManager.java
package com.spvessel.spacevil;

import com.spvessel.spacevil.Core.IEventUpdate;
import com.spvessel.spacevil.Flags.GeometryEventType;

import java.util.Map;
import java.util.HashMap;
import java.util.LinkedHashSet;

class EventManager {
    //protected static Boolean isLocked = true;
    private Map<GeometryEventType, LinkedHashSet<IEventUpdate>> listeners = new HashMap<>();

    void subscribe(GeometryEventType type, IEventUpdate listener) {
        if (!listeners.containsKey(type))
            listeners.put(type, new LinkedHashSet<IEventUpdate>());

        if (!listeners.get(type).contains(listener))
            listeners.get(type).add(listener);
    }

    void unsubscribe(GeometryEventType type, IEventUpdate listener) {
        if (listeners.containsKey(type)) {
            if (listeners.get(type).contains(listener)) {
                listeners.get(type).remove(listener);
            }
        }
    }

    void notifyListeners(GeometryEventType type, int value) {
        if (listeners.containsKey(type)) {
            for (IEventUpdate item : listeners.get(type)) {
                item.update(type, value);
            }
        }
    }
}

File: SpaceVIL/Core/EventManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SpaceVIL.Core
{
    internal sealed class EventManager
    {
        //internal static bool IsLocked = true;
        internal Dictionary<GeometryEventType, LinkedHashSet<IEventUpdate>> listeners = new Dictionary<GeometryEventType, LinkedHashSet<IEventUpdate>>();

        internal void Subscribe(GeometryEventType type, IEventUpdate listener)
        {
            if (!listeners.ContainsKey(type))
                listeners.Add(type, new LinkedHashSet<IEventUpdate>());

            if (!listeners[type].Contains(listener))
                listeners[type].Add(listener);
        }

        internal void Unsubscribe(GeometryEventType type, IEventUpdate listener)
        {
            if (listeners.ContainsKey(type))
                if (listeners[type].Contains(listener))
                    listeners[type].Remove(listener);
        }

        internal void NotifyListeners(GeometryEventType type, int value)
        {
            if (listeners.ContainsKey(type))
                foreach (var _ in listeners[type])
                    _.Update(type, value);
        }
    }
}
